src/webots/nodes/WbBaseNode.cpp

```diff
@@ -124,6 +125,269 @@ void WbBaseNode::validateProtoNodes() {
+  const QList<WbNode *> nodes = subNodes(true, true, true);
+
+  // the internal node is used to keep track of what can be collapsed since it's the bottom of the chain and it's unique
+  // whereas the chain itself can be comprised of multiple parameter nodes which complicates keeping track of how they relate
+  QList<WbNode *> internalProtoNodes;
+
+  for (int i = 0; i < nodes.size(); ++i)
+    if (nodes[i]->isInternalNode())
+      internalProtoNodes.append(nodes[i]);
+  /*
+  // TODO: remove before merge
+  printf("PRINT CHAINS FOR UNFILTERED CANDIDATES\n");
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    printf("\n");
+    printChainCandidate(internalProtoNodes[i]);
+  }
+  */
+
+  QList<WbNode *> tmp = internalProtoNodes;
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    if (isInternalNodeVisible(internalProtoNodes[i])) {
+      // cannot collapse visible nodes otherwise they no longer refresh on the interface
+      tmp.removeAll(internalProtoNodes[i]);
+      // also remove among the candidates any ancestor to this node otherwise it will be deleted indirectly.
+      // likewise any descendants can't be deleted either as they might be referenced indirectly (e.g if the texture url field
+      // is visible, the corresponding TextureCoordinate/IndexedFaceSet nodes can't be deleted even if themselves aren't)
+      for (int j = 0; j < internalProtoNodes.size(); ++j)
+        if (internalProtoNodes[j]->isAnAncestorOf(internalProtoNodes[i]) ||
+            internalProtoNodes[i]->isAnAncestorOf(internalProtoNodes[j]))
+          tmp.removeAll(internalProtoNodes[j]);
+    }
+  }
+  internalProtoNodes = tmp;
+
+  QList<WbNode *> invisibleProtoParameterNodes;
+  // follow the chain upwards, starting from the internal node, to extract all the protoParameterNodes that can be deleted
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    WbNode *n = internalProtoNodes[i]->protoParameterNode();
+
+    while (n != NULL) {
+      bool added = false;
+      for (int j = 0; j < invisibleProtoParameterNodes.size(); ++j)
+        if (n->level() > invisibleProtoParameterNodes[j]->level()) {  // insert them from lowest to highest level
+          invisibleProtoParameterNodes.insert(j, n);
+          added = true;
+          break;  // need to break otherwise invisibleProtoParameterNodes grows infinitly
+        }
+
+      if (!added)
+        invisibleProtoParameterNodes.append(n);
+
+      n = n->protoParameterNode();
+    }
+  }
+  /*
+  // TODO: remove before merge
+  printf("\nINVISIBLE PROTO PARAMETER NODES (WHAT WILL BE REMOVED)\n");
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    printf("  [L%d] %s [%p]\n", invisibleProtoParameterNodes[i]->level(),
+           invisibleProtoParameterNodes[i]->usefulName().toUtf8().constData(), invisibleProtoParameterNodes[i]);
+  }
+  */
+
+  if (invisibleProtoParameterNodes.size() == 0)
```

Comment:
usually we use the isEmpty() method
```suggestion
  if (invisibleProtoParameterNodes.isEmpty())
```